scripts/generate_docs.sc

```diff
@@ -0,0 +1,157 @@
+    // to avoid errors during branch checkout
+    cp.over(pwd/"generate_docs.sc", pwd/"generate_docs.temp")
+    %('amm, "generate_docs.temp", "--tracked_by_git", "false")
+    rm! pwd/"generate_docs.temp"
+  } else {
+    //check that repo is clean
+    val possiblyChangedLines: Vector[String] = %%('git, "diff-files", "--name-only").out.lines
+    if (possiblyChangedLines.nonEmpty) {
+      val msg = s"Git repository isn't clean, aborting docs generation. Changed files:\n${possiblyChangedLines.mkString("\n")}"
+      println(msg)
+      exit()
+    }
+
+    val buildDir = makeTmpDir()
+    val docsDir = pwd/up/"docs"
+
+    buildDocs(buildDir, docsDir)
+    %('git, 'checkout, currentGitBranch)
+    println(s"Success! Docs were generated at ${buildDir/'docs/'_site}")
+  }
+}
+
+def makeTmpDir(): Path = {
+  val timestamp = Instant.now().toString.replace(':', '-')
+  val path = root/"tmp"/s"marathon-docs-build-$timestamp"
+  mkdir! path
+  path
+}
+
+val ignoredReleaseBranchesVersions = Set("1.6")
+
+val currentGitBranch = %%('git, "rev-parse", "--abbrev-ref", "HEAD").out.lines.head
+
+def listAllTagsInOrder: Vector[String] = %%('git, "tag", "-l", "--sort=version:refname").out.lines
+
+def isReleaseTag(tag: String) = tag.matches("""v[1-9]+\.\d+\.\d+""")
+
+/**
+  * get the major.minor version from a [v]major.minor.patch
+  *
+  * v1.4.11 -> 1.4
+  * 1.45.983 -> 1.45
+  *
+  * @param tag
+  * @return
+  */
+def releaseBranchVersion(tag: String) = tag.dropWhile(_.isLetter).takeWhile {
+  var reachedMinorVersion = false
+  elem => elem match {
+    case c if c.isDigit => true
+    case c if c == '.' && !reachedMinorVersion =>
+      reachedMinorVersion = true
+      true
+    case _ => false
+  }
+}
+
+def notIgnoredBranch(branchVersion: String) = !ignoredReleaseBranchesVersions.contains(branchVersion)
+
+def getTheLastTagVersion(tags: Seq[String]) = tags.last.drop(1)
+
+/**
+  * getting 3 latest release versions and corresponding latest tags
+  *
+  * example: List[(String, String)] = List(("1.3", "1.3.14"), ("1.4", "1.4.11"), ("1.5", "1.5.6"))
+  */
+val docsTargetVersions = listAllTagsInOrder
+  .filter(isReleaseTag)
+  .groupBy(releaseBranchVersion)
+  .filterKeys(notIgnoredBranch)
+  .mapValues(getTheLastTagVersion)
+  .toList.sortBy(_._1)
+  .takeRight(3)
+
+val (latestReleaseBranch, latestReleaseVersion) = docsTargetVersions.last
+
+def generateDocsForVersion(docsPath: Path, version: String, outputPath: String = "_site"): Unit = {
+  %('bundle, "exec", s"jekyll build --config _config.yml,_config.$version.yml -d $outputPath/$version/")(docsPath)
+}
+
+def branchForTag(version: String) = {
+  s"tags/v$version"
+}
+
+// step 1: copy docs/docs to the respective dirs
+
+def checkoutDocsToTempFolder(buildDir: Path, docsDir: Path) = {
+  docsTargetVersions foreach { case (releaseBranchVersion, tagVersion) =>
+    val tagName = branchForTag(tagVersion)
+    %git('checkout, s"$tagName")
+    val tagBuildDir = buildDir / releaseBranchVersion
```

Comment:
How about returning this directory so you can use the value in `generateVersionedDocs` and not rely on convention?

Reply:
We can do it, but I'm not sure if it will make the code cleaner.

Reply:
IMHO this is the build artifact of this method. What is the benefit of relying on convention here when the compiler can ensure that the correct dir is used?